### PR TITLE
Accept custom reporter as constructor

### PR DESCRIPTION
**Pull request by just-boris**
Example with the custom reporter doesn't work in the branch `4.0`.

I am going to use it to test my Allure integration with Webdriver.io, so it is matters for me.

So, I've fixed it.

### CODE REVIEW

**Comment by christian-bromann** [comment]
Why does the current code not work for you?

**Comment by just-boris** [comment]
This configuration from example doesn't work: https://github.com/webdriverio/webdriverio/blob/master/examples/wdio/wdio.customreporter.conf.js#L39 

It always loads npm-module, but I want to pass function here.

**Comment by christian-bromann** [comment]
How about changing that line to
```js
reporter: [__dirname + ''./reporter/my.custom.reporter'']
```
or something like this

**Comment by just-boris** [comment]
It doesn't work too. Reporter name is always embraced as `wdio-${reporterType}-reporter`

File: lib/launcher.js [hunk 1]
```diff
@@ -47,13 +47,15 @@ class Launcher {
...
-                throw new Error(`reporter "wdio-${reporterType}-reporter" is not installed. Error: ${e.stack}`)
+        for (let Reporter of config.reporter) {
+            if (typeof Reporter === 'string') {
+                try {
+                    Reporter = require(`wdio-${Reporter}-reporter`)
+                } catch (e) {
+                    throw new Error(`reporter "wdio-${Reporter}-reporter" is not installed. Error: ${e.stack}`)
+                }
             }
+            reporter.add(new Reporter(reporter, config))
```
Comment (christian-bromann) [inline lib/launcher.js]: I don't think this is the right way we should implement this. What if `Reporter` has a spelling issue and is still a string? How about we do an `else if` and check if typeof `reporterType` is function?

**Comment by just-boris** [comment]
Fixed that.
If reporter is not installed or has a spelling issue, the error will be thrown.
If the option has unknown type, error will be thrown too. Checked that on my local setup.

File: lib/launcher.js [hunk 2]
```diff
@@ -47,13 +47,18 @@ class Launcher {
...
+        for (let Reporter of config.reporter) {
+            if (typeof Reporter === 'string') {
+                try {
+                    Reporter = require(`wdio-${Reporter}-reporter`)
+                } catch (e) {
+                    throw new Error(`reporter "wdio-${Reporter}-reporter" is not installed. Error: ${e.stack}`)
+                }
             }
+            if (typeof Reporter !== 'function') {
+                throw new Error(`reporter option must be either string or function, but got: ${Reporter}`)
```
Comment (christian-bromann) [inline lib/launcher.js]: What do you think about: `...string or function, but got: ${typeof Reporter}')`?

Comment (just-boris) [inline lib/launcher.js]: `typeof null == 'object'`, so sometimes that message can not be informative.

Comment (just-boris) [inline lib/launcher.js]: Usually, most values are converted to string as `[object Object]` and I considered that is would be good to explain the error.

Comment (christian-bromann) [inline lib/launcher.js]: true.

**Comment by christian-bromann** [comment]
thanks @just-boris

**Pull request merged**